File: patterns/bridge.js
//bridge pattern
//link what it is to how it works

// Define the Abstraction interface, which provides high-level functionality to the client
class Abstraction {
    constructor(implementation) {
        this.implementation = implementation;
    }

    operation() {
        const implementationResult = this.implementation.operationImplementation();
        console.log(`Abstraction: Base operation with implementation result: ${implementationResult}`);
    }
}

// Define the Implementor interface, which provides low-level functionality that Abstraction can rely on
class Implementor {
    operationImplementation() { }
}

// Define Concrete Implementors, which provide specific implementations of the Implementor interface
class ConcreteImplementorA extends Implementor {
    operationImplementation() {
        return "ConcreteImplementorA: Operation";
    }
}

class ConcreteImplementorB extends Implementor {
    operationImplementation() {
        return "ConcreteImplementorB: Operation";
    }
}

// Use the Abstraction and Implementor together to provide different combinations of functionality
const implementationA = new ConcreteImplementorA();
const abstractionA = new Abstraction(implementationA);
abstractionA.operation();

const implementationB = new ConcreteImplementorB();
const abstractionB = new Abstraction(implementationB);
abstractionB.operation();

File: patterns/bridge.py
# Define the Abstraction class, which defines the high-level interface that the client uses
class Abstraction:
    def __init__(self, implementation):
        self._implementation = implementation

    def operation(self):
        return f"Abstraction: {self._implementation.operation_implementation()}"

# Define the Implementation interface, which defines the low-level interface for the concrete implementations
class Implementation:
    def operation_implementation(self):
        pass

# Define the ConcreteImplementation classes, which implement the Implementation interface
class ConcreteImplementationA(Implementation):
    def operation_implementation(self):
        return "ConcreteImplementationA: Doing operation"

class ConcreteImplementationB(Implementation):
    def operation_implementation(self):
        return "ConcreteImplementationB: Doing operation"

# Use the Abstraction and Implementation interfaces to create a decoupled hierarchy
implementation_a = ConcreteImplementationA()
abstraction_a = Abstraction(implementation_a)

implementation_b = ConcreteImplementationB()
abstraction_b = Abstraction(implementation_b)

print(abstraction_a.operation())
print(abstraction_b.operation())
